From 3bf2da6a4243bb361f69b0d6a886ef5bd18d6fd9 Mon Sep 17 00:00:00 2001
From: Sergey Fedorov <barracuda@macos-powerpc.org>
Date: Fri, 7 Aug 2026 03:16:52 +0000
Subject: [PATCH 18/38] cursor (Apple): use dictionaryWithContentsOfFile and
 valueForKey

Avoids the NSURL/error-based plist loader and Objective-C subscripting
literal syntax that isn't available on older toolchains; wrap in an
NSAutoreleasePool on pre-10.9 targets.

Ported from https://github.com/macos-powerpc/powerpc-ports/blob/main/sysutils/fastfetch/files/0012-Misc-fixes-for-legacy-systems.patch
---
 src/detection/cursor/cursor_apple.m | 37 ++++++++++++++++-------------
 1 file changed, 20 insertions(+), 17 deletions(-)

diff --git a/src/detection/cursor/cursor_apple.m b/src/detection/cursor/cursor_apple.m
index 610bd1ed8..a5eae615d 100644
--- a/src/detection/cursor/cursor_apple.m
+++ b/src/detection/cursor/cursor_apple.m
@@ -1,13 +1,22 @@
 #include "cursor.h"
 
-#import <Foundation/Foundation.h>
+#include <Foundation/Foundation.h>
+#include <AvailabilityMacros.h>
+
+#if MAC_OS_X_VERSION_MIN_REQUIRED < 1090
+#define POOLSTART NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
+#define POOLEND   [pool release];
+#else
+#define POOLSTART
+#define POOLEND
+#endif
 
 static void appendColor(FFstrbuf* str, NSDictionary* dict)
 {
-    uint32_t r = (uint32_t) (((NSNumber*) dict[@"red"]).doubleValue * 255 + .5);
-    uint32_t g = (uint32_t) (((NSNumber*) dict[@"green"]).doubleValue * 255 + .5);
-    uint32_t b = (uint32_t) (((NSNumber*) dict[@"blue"]).doubleValue * 255 + .5);
-    uint32_t a = (uint32_t) (((NSNumber*) dict[@"alpha"]).doubleValue * 255 + .5);
+    uint32_t r = (uint32_t) (((NSNumber*) [dict valueForKey:@"red"]).doubleValue * 255 + .5);
+    uint32_t g = (uint32_t) (((NSNumber*) [dict valueForKey:@"green"]).doubleValue * 255 + .5);
+    uint32_t b = (uint32_t) (((NSNumber*) [dict valueForKey:@"blue"]).doubleValue * 255 + .5);
+    uint32_t a = (uint32_t) (((NSNumber*) [dict valueForKey:@"alpha"]).doubleValue * 255 + .5);
     uint32_t color = (r << 24) | (g << 16) | (b << 8) | a;
 
     switch (color)
@@ -30,34 +39,28 @@ static void appendColor(FFstrbuf* str, NSDictionary* dict)
 
 void ffDetectCursor(FFCursorResult* result)
 {
-    NSError* error;
-    NSString* fileName = [NSString stringWithFormat:@"file://%s/Library/Preferences/com.apple.universalaccess.plist", instance.state.platform.homeDir.chars];
-    NSDictionary* dict = [NSDictionary dictionaryWithContentsOfURL:[NSURL URLWithString:fileName]
-                                       error:&error];
-    if(error)
-    {
-        ffStrbufAppendS(&result->error, error.localizedDescription.UTF8String);
-        return;
-    }
+    POOLSTART
+    NSDictionary* dict = [NSDictionary dictionaryWithContentsOfFile:[NSHomeDirectory() stringByAppendingPathComponent:@"/Library/Preferences/com.apple.universalaccess.plist"]];
 
     NSDictionary* color;
 
     ffStrbufAppendS(&result->theme, "Fill - ");
-    if ((color = dict[@"cursorFill"]))
+    if ((color = [dict valueForKey:@"cursorFill"]))
         appendColor(&result->theme, color);
     else
         ffStrbufAppendS(&result->theme, "Black");
 
     ffStrbufAppendS(&result->theme, ", Outline - ");
 
-    if ((color = dict[@"cursorOutline"]))
+    if ((color = [dict valueForKey:@"cursorOutline"]))
         appendColor(&result->theme, color);
     else
         ffStrbufAppendS(&result->theme, "White");
 
-    NSNumber* mouseDriverCursorSize = dict[@"mouseDriverCursorSize"];
+    NSNumber* mouseDriverCursorSize = [dict valueForKey:@"mouseDriverCursorSize"];
     if (mouseDriverCursorSize)
         ffStrbufAppendF(&result->size, "%d", (int) (mouseDriverCursorSize.doubleValue * 32 + 0.5));
     else
         ffStrbufAppendS(&result->size, "32");
+    POOLEND
 }
-- 
2.43.0

